docs/cudf/source/developer_guide/ARCHITECTURE.md

```diff
@@ -0,0 +1,208 @@
+`BaseIndex` is generally intended to be a true abstract class, i.e. it should contain no implementations.
+Functions may be implemented in `BaseIndex` if they are truly identical for all types of indexes.
+However, currently most such implementations are not applicable to all subclasses and will be eventually be removed.
+
+Almost all indexes are subclasses of `GenericIndex`, a single-columned index with the class hierarchy:
+```python
+class GenericIndex(SingleColumnFrame, BaseIndex)
+```
+`Frame`->`SingleColumnFrame`->`GenericIndex`<-`BaseIndex`.
+Integer, float, or string indexes are all composed of a single column of data.
+Most `GenericIndex` methods are inherited from `Frame`, saving us the trouble of rewriting them.
+
+We now consider the three main exceptions to this model:
+
+- A `RangeIndex` is not backed by a column of data, so it inherits directly from `BaseIndex` alone.
+  Wherever possible, its methods have special implementations designed to avoid materializing columns.
+  Where such an implementation is infeasible, we fall back to converting it to an integer index first instead.
+- A `MultiIndex` is backed by _multiple_ columns of data.
+  Therefore, its inheritance hierarchy looks like `class MultiIndex(Frame, BaseIndex)`.
+  Some of its more `Frame`-like methods may be inherited,
+  but many others must be reimplemented since in many cases a `MultiIndex` is not expected to behave like a `Frame`.
+- Just like in pandas, `Index` itself can never be instantiated.
+  `pandas.Index` is the parent class for indexes,
+  but its constructor returns an appropriate subclass depending on the input data type and shape.
+  Unfortunately, mimicking this behavior requires overriding `__new__`,
+  which in turn makes shared initialization across inheritance trees much more cumbersome to manage.
+  To enable sharing constructor logic across different index classes,
+  we instead define `BaseIndex` as the parent class of all indexes.
+  `Index` inherits from `BaseIndex`, but it masquerades as a `BaseIndex` to match pandas.
+  This class should contain no implementations since it is simply a factory for other indexes.
+
+
+## The Column layer
+
+The next layer in the cuDF stack is the Column layer.
+This layer forms the glue between pandas-like APIs and our underlying data layouts.
+Under the hood, cuDF is built around the [Apache Arrow Format](https://arrow.apache.org).
+This data format is both conducive to high-performance algorithms and suitable for data interchange between libraries.
+
+The principal objects in the Column layer are the `ColumnAccessor` and the various `Column` classes.
+The `Column` is cuDF's core data structure that represents a single column of data of a specific data type.
+A `ColumnAccessor` is a dictionary-like interface to a sequence of `Column`s.
+A `Frame` owns a `ColumnAccessor`, and most of its operations are implemented as loops over that object's `Column`s.
```

Comment:
I don't know if we want to paint this picture as hallmark algorithms like I/O, `join` and `groupby` don't work this way.

Reply:
Yeah, this just sounds inefficient. I forgot to comment on this earlier but it raised an eyebrow for me.

Reply:
I can remove this sentence, but it's important to highlight the reason that I wrote this because it's an area for improvement (not just in this document, but in our actual design).

We have to distinguish between methods that act on a per-column basis and methods that act on the table as a whole. The former generally _do_ follow this pattern (and it's not inefficient, or at least no more so than any alternative in a pure Python world), while the latter do not. This distinction is also where the Frame/Column layer boundary breaks down, because in practice any APIs that act on the entire table bypass the Column layer and go straight to libcudf. I don't want to try and address this discrepancy in the initial version of this document, but we may need to do so eventually. At that point we may want to come back and make more broad statements characterizing how different kinds of APIs are implemented.

Reply:
Actually I think there is a more natural way to incorporate this later on in the document.